main/lsp/LSPTypechecker.cc

```diff
@@ -335,19 +342,26 @@ LSPTypechecker::SlowPathResult LSPTypechecker::runSlowPath(LSPFileUpdates &updat
     }
     logger->debug("Taking slow path");
 
-    auto finalGS = move(updates.updatedGS.value());
+    ENFORCE(this->cancellationUndoState == nullptr);
+    if (cancelable) {
+        this->cancellationUndoState =
+            std::make_unique<UndoState>(std::move(gs), std::move(this->indexedFinalGS), updates.epoch);
+    }
+
+    this->gs = std::move(updates.updatedGS.value());
+
     const uint32_t epoch = updates.epoch;
-    auto &epochManager = *finalGS->epochManager;
+    auto &epochManager = *this->gs->epochManager;
     // Note: Commits can only be canceled if this edit is cancelable, LSP is running across multiple threads, and the
     // cancelation feature is enabled.
-    auto committed = epochManager.tryCommitEpoch(*finalGS, epoch, cancelable, preemptManager, [&]() -> void {
+    auto committed = epochManager.tryCommitEpoch(*this->gs, epoch, cancelable, preemptManager, [&]() -> void {
         vector<ast::ParsedFile> indexed, nonPackagedIndexed;
 
         {
             // Replace error queue with one that is owned by this thread.
```

Comment:
If we stop initializing `this->gs` with `updates.updatedGS`, we can drop this `exchange`, right?